src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,139 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<T>(null, value, null);
+        if (head == null) {
+            head = node;
+        } else {
+            node.previous = tail;
+            tail.next = node;
+        }
+        tail = node;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            head.previous = new Node<>(null, value, head);
+            head = head.previous;
+        } else {
+            Node<T> validNode = getNodeByIndex(index);
+            Node<T> node = new Node<T>(validNode.previous, value, validNode);
+            validNode.previous = node;
+            node.next.previous = node;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNodeByIndex(index);
+        T oldVal = node.value;
+        node.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removeTarget = getNodeByIndex(index);
+        unlink(removeTarget);
+        return removeTarget.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = getNodeByValue(object);
+        if (node == null) {
+            return false;
+        }
+        unlink(node);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> previous;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void throwException(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index out of bounds for length");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int targetIndex) {
+        throwException(targetIndex);
+        Node<T> currentNode = tail;
+        for (int i = size - 1; i != targetIndex; i--) {
+            currentNode = currentNode.previous;
+        }
+        return currentNode;
+    }
```

Comment:
Imagine you heve 10 000 elemenst and you need to find  3-rd element. In that case, you will need 9 997 iterations to find  3 element. Youn need to start searching the element from head or tail, depending on index.